spec/android_unit_test_pr_checker_spec.rb

```diff
@@ -0,0 +1,222 @@
+          'Abc.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { System.out.println(""); } }',
+          'project/src/androidTest/java/org/test/AbcTests.java' => 'class AbcTests { public void abcTest() { sut.main([]) } }',
+          'Polygon.kt' => 'abstract class Polygon { abstract fun draw() }',
+          'project/src/androidTest/java/org/test/PolygonTest.kt' => 'class PolygonTest { fun drawTest() {}',
+          'TestsINeedThem.java' => 'public final class TestsINeedThem { public void testMe() { System.out.println(""); } }',
+          'project/src/androidTest/java/org/test/TestsINeedThemTests.java' => 'class TestsINeedThemTests { public void testMe() {} }',
+          'MyNewClass.java' => 'final class MyNewClass { public void testMe() { System.out.println(""); } }',
+          'project/src/androidTest/java/org/test/MyNewClass.java' => 'class MyNewClass { public void testMe() {} }'
+        }
+
+        diff = generate_add_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+
+      it 'does not show errors when we are deleting classes' do
+        changes_dict = {
+          'Abc.java' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { System.out.println(""); } }',
+          'Polygon.kt' => 'abstract class Polygon { abstract fun draw() }',
+          'TestsINeedThem.java' => 'public final class TestsINeedThem { public void testMe() { System.out.println(""); } }'
+        }
+
+        diff = generate_deleted_diff(changes_dict)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests
+
+        expect(@dangerfile.status_report[:errors]).to be_empty
+      end
+
+      it 'show errors when we remove test classes for classes we refactored' do
+        added_classes = {
+          'Abc.kt' => 'import java.utils.*;\n\n public class Abc { public static void main(String[] args) { System.out.println(""); } }',
+          'Polygon.kt' => 'data class Polygon(sides: Int) { fun draw() {} }',
+          'TestsINeedThem.kt' => 'public open class TestsINeedThem { fun testMe2() { } }'
+        }
+
+        removed_tests = {
+          'project/src/androidTest/java/org/test/AbcTests.java' => 'class AbcTests { public void abcTest() { sut.main([]) } }',
+          'project/src/androidTest/java/org/test/PolygonTest.java' => 'class PolygonTest { void drawTest() {}',
+          'project/src/androidTest/java/org/test/TestsINeedThem.java' => 'class TestsINeedThem { public void testMe() {} }'
+        }
+
+        diff = generate_add_diff(added_classes) + generate_deleted_diff(removed_tests)
+        allow(@dangerfile.git).to receive(:diff).and_return(diff)
+
+        @plugin.check_missing_tests
+
+        expect(@dangerfile.status_report[:errors].count).to eq 3
```

Comment:
I'd suggest your Unit Tests to not just verify if the number of errors is the expected one, but to also check if the error messages that are reported are the expected one.

Because if we change the implementation one day in some breaking way, but by sheer change that makes one violation that was triggered before not be triggered anymore… while it also makes another violation that was not triggered before now be triggered… you'd end up with a `n+1-1` situation that would make your errors count unchanged and would have your spec still pass, making you think it still works while your refactoring of the implementation in fact broke things.